File: tools/command/disassemble.cpp
#include <fmt/base.h>
#include <fmt/format.h>

#include <libxdb/disassembler.hpp>
#include <libxdb/parse.hpp>
#include <libxdb/types.hpp>

#include "handlers.hpp"

namespace xdb_handlers {

void print_disassembly(xdb::process &process, xdb::virt_addr address, std::size_t n_instructions) {
    xdb::disassembler dis(process);
    auto instructions = dis.disassemble(n_instructions, address);
    for (const auto &instruction : instructions) {
        fmt::println("{:08x}: {}", instruction.address.addr(), instruction.text);
    }
}

void handle_disassemble_command(xdb::process &process, std::span<const std::string> args) {
    auto address = process.get_pc();
    constexpr std::size_t default_instruction_count = 5;
    std::size_t n_instructions = default_instruction_count;

    // Parse command line arguments for -c and -a flags
    for (std::size_t i = 1; i < args.size(); ++i) {
        if (args[i] == "-c" && i + 1 < args.size()) {
            auto count = xdb::to_integral<std::size_t>(args[i + 1]);
            if (!count) {
                fmt::println("Invalid instruction count format.");
                return;
            }
            n_instructions = *count;
            ++i;  // Skip the argument we just consumed
        } else if (args[i] == "-a" && i + 1 < args.size()) {
            constexpr int hex_base = 16;
            auto addr = xdb::to_integral<std::uint64_t>(args[i + 1], hex_base);
            if (!addr) {
                fmt::println("Invalid address format. Address is expected in 0xhex format.");
                return;
            }
            address = xdb::virt_addr{*addr};
            ++i;  // Skip the argument we just consumed
        }
    }

    print_disassembly(process, address, n_instructions);
}

}  // namespace xdb_handlers

File: src/bin/xdb/commands/disassemble.rs
use xdb::{Disassembler, Process, VirtAddr, parse};

const DEFAULT_INSTRUCTION_COUNT: usize = 5;

pub fn handle(process: &mut Process, args: &[&str]) -> xdb::Result<()> {
    let mut address = None;
    let mut count = DEFAULT_INSTRUCTION_COUNT;

    // Parse command line arguments for -c and -a flags
    let mut i = 1;
    while i < args.len() {
        match args[i] {
            "-c" if i + 1 < args.len() => {
                let Some(n) = parse::to_integral::<usize>(args[i + 1], 10) else {
                    eprintln!("Invalid instruction count.");
                    return Ok(());
                };
                count = n;
                i += 2; // Skip the argument we just consumed
            }
            "-a" if i + 1 < args.len() => {
                let Some(addr) = parse::to_integral::<u64>(args[i + 1], 16) else {
                    eprintln!("Invalid address. Use 0x prefix for hex addresses.");
                    return Ok(());
                };
                address = Some(VirtAddr(addr));
                i += 2; // Skip the argument we just consumed
            }
            other => {
                eprintln!("Unknown argument: {other}");
                return Ok(());
            }
        }
    }

    print_disassembly(process, address, count)
}

pub fn print_disassembly(
    process: &Process,
    address: Option<VirtAddr>,
    count: usize,
) -> xdb::Result<()> {
    let instructions = Disassembler::new(process).disassemble(count, address)?;
    for instruction in instructions {
        println!("{:#018x}: {}", instruction.address.addr(), instruction.text);
    }
    Ok(())
}
